### PR TITLE
Server/Agent websocket client reconnect backoff

### LINKED ISSUE #2027: Agent reconnect exponential backoff
When a Kontena master is down but the HAproxy instance is not, the agents try to reconnect once per second indefinitely. It would be nice if this was configurable or if there was a back-off algorithm in place so that a master that is down for a long time is not causing unnecessary resource utilization. For example, with 7 nodes when my master is down there is a constant ~68KiB/s in and out. With 100 nodes that would be about 7.5Mbit/s just for reconnection attempts from all of the nodes and 

I propose an exponential backoff to a random number between 45 and 75 seconds between attempts (random number chosen again after each attempt).

### PR DESCRIPTION
**Pull request by SpComb**
Fixes #2027 for both agent -> server and server -> cloud

Replace the `start` -> `every(...)` -> `connect unless connecting?` state with explicit `start/reconnect` -> `after(...)` -> `connect!` timers.

Connect errors or websocket closes will reconnect with increasing backoff, and websocket opens will reset the reconnect backoff.

The reconnect interval uses capped exponential backoff with [randomized full jitter](https://www.awsarchitectureblog.com/2015/03/backoff.html) to spread out the incoming server/cloud websocket connections.

### Server
```
I, [2017-10-06T10:08:43.568184 #9]  INFO -- Cloud::WebsocketClient: Connecting to cloud at wss://socket.kontena.io/platform
I, [2017-10-06T10:08:44.729585 #9]  INFO -- Cloud::WebsocketClient: Connected to wss://socket.kontena.io/platform with ssl verify: /OU=Domain Control Validated/OU=PositiveSSL Wildcard/CN=*.kontena.io (issuer /C=GB/ST=Greater Manchester/L=Salford/O=COMODO CA Limited/CN=COMODO RSA Domain Validation Secure Server CA)
...
I, [2017-10-06T10:09:48.609163 #9]  INFO -- Cloud::WebsocketClient: cloud connection closed with code 1006: read timeout after 4.999060717s while waiting 5.0s for pong
I, [2017-10-06T10:09:48.612059 #9]  INFO -- Cloud::WebsocketClient: reconnect attempt 1 in 3.72s...
I, [2017-10-06T10:09:52.337178 #9]  INFO -- Cloud::WebsocketClient: Connecting to cloud at wss://socket.kontena.io/platform
I, [2017-10-06T10:10:05.664215 #9]  INFO -- Cloud::WebsocketClient: cloud connection closed with code 1006: Host is unreachable - connect(2) for 54.227.177.16:443
I, [2017-10-06T10:10:05.689120 #9]  INFO -- Cloud::WebsocketClient: reconnect attempt 2 in 6.59s...
I, [2017-10-06T10:10:12.278510 #9]  INFO -- Cloud::WebsocketClient: Connecting to cloud at wss://socket.kontena.io/platform
I, [2017-10-06T10:10:23.583045 #9]  INFO -- Cloud::WebsocketClient: cloud connection closed with code 1006: Host is unreachable - connect(2) for 54.227.177.16:443
I, [2017-10-06T10:10:23.584012 #9]  INFO -- Cloud::WebsocketClient: reconnect attempt 3 in 14.93s...
I, [2017-10-06T10:10:38.513631 #9]  INFO -- Cloud::WebsocketClient: Connecting to cloud at wss://socket.kontena.io/platform
I, [2017-10-06T10:10:45.664261 #9]  INFO -- Cloud::WebsocketClient: cloud connection closed with code 1006: Host is unreachable - connect(2) for 54.227.177.16:443
I, [2017-10-06T10:10:45.689159 #9]  INFO -- Cloud::WebsocketClient: reconnect attempt 4 in 16.58s...
I, [2017-10-06T10:11:02.268151 #9]  INFO -- Cloud::WebsocketClient: Connecting to cloud at wss://socket.kontena.io/platform
I, [2017-10-06T10:11:09.376354 #9]  INFO -- Cloud::WebsocketClient: cloud connection closed with code 1006: Host is unreachable - connect(2) for 54.224.148.229:443
I, [2017-10-06T10:11:09.379100 #9]  INFO -- Cloud::WebsocketClient: reconnect attempt 5 in 37.79s...
I, [2017-10-06T10:11:47.168094 #9]  INFO -- Cloud::WebsocketClient: Connecting to cloud at wss://socket.kontena.io/platform
I, [2017-10-06T10:11:54.302606 #9]  INFO -- Cloud::WebsocketClient: cloud connection closed with code 1006: Host is unreachable - connect(2) for 54.224.148.229:443
I, [2017-10-06T10:11:54.303305 #9]  INFO -- Cloud::WebsocketClient: reconnect attempt 6 in 64.60s...
I, [2017-10-06T10:12:58.908079 #9]  INFO -- Cloud::WebsocketClient: Connecting to cloud at wss://socket.kontena.io/platform
I, [2017-10-06T10:13:06.047806 #9]  INFO -- Cloud::WebsocketClient: cloud connection closed with code 1006: Host is unreachable - connect(2) for 54.227.177.16:443
I, [2017-10-06T10:13:06.050075 #9]  INFO -- Cloud::WebsocketClient: reconnect attempt 7 in 35.91s...
I, [2017-10-06T10:13:41.957223 #9]  INFO -- Cloud::WebsocketClient: Connecting to cloud at wss://socket.kontena.io/platform
I, [2017-10-06T10:13:49.087537 #9]  INFO -- Cloud::WebsocketClient: cloud connection closed with code 1006: Host is unreachable - connect(2) for 54.224.148.229:443
I, [2017-10-06T10:13:49.090119 #9]  INFO -- Cloud::WebsocketClient: reconnect attempt 8 in 7.26s...
I, [2017-10-06T10:13:56.349092 #9]  INFO -- Cloud::WebsocketClient: Connecting to cloud at wss://socket.kontena.io/platform
I, [2017-10-06T10:14:03.487285 #9]  INFO -- Cloud::WebsocketClient: cloud connection closed with code 1006: Host is unreachable - connect(2) for 54.224.148.229:443
I, [2017-10-06T10:14:03.488555 #9]  INFO -- Cloud::WebsocketClient: reconnect attempt 9 in 83.35s...
```

### Agent
```
D, [2017-10-06T10:16:43.433453 #1] DEBUG -- Kontena::WebsocketClient: Server closed connection with code 1006: Server closed connection without sending close frame
W, [2017-10-06T10:16:43.433674 #1]  WARN -- Kontena::WebsocketClient: connection closed with code 1006: Server closed connection without sending close frame
I, [2017-10-06T10:16:43.434086 #1]  INFO -- Kontena::WebsocketClient: reconnect attempt 1 in 0.91s...
I, [2017-10-06T10:16:44.349730 #1]  INFO -- Kontena::WebsocketClient: connecting to master at ws://192.168.66.1:9292
E, [2017-10-06T10:16:44.360665 #1] ERROR -- Kontena::WebsocketClient: unable to connect to server: Connection refused - connect(2) for 192.168.66.1:9292
I, [2017-10-06T10:16:44.362733 #1]  INFO -- Kontena::WebsocketClient: reconnect attempt 2 in 0.00s...
I, [2017-10-06T10:16:44.369617 #1]  INFO -- Kontena::WebsocketClient: connecting to master at ws://192.168.66.1:9292
E, [2017-10-06T10:16:44.379800 #1] ERROR -- Kontena::WebsocketClient: unable to connect to server: Connection refused - connect(2) for 192.168.66.1:9292
I, [2017-10-06T10:16:44.381855 #1]  INFO -- Kontena::WebsocketClient: reconnect attempt 3 in 3.11s...
I, [2017-10-06T10:16:47.497299 #1]  INFO -- Kontena::WebsocketClient: connecting to master at ws://192.168.66.1:9292
E, [2017-10-06T10:16:47.499785 #1] ERROR -- Kontena::WebsocketClient: unable to connect to server: Connection refused - connect(2) for 192.168.66.1:9292
I, [2017-10-06T10:16:47.500335 #1]  INFO -- Kontena::WebsocketClient: reconnect attempt 4 in 3.03s...
I, [2017-10-06T10:16:50.528600 #1]  INFO -- Kontena::WebsocketClient: connecting to master at ws://192.168.66.1:9292
E, [2017-10-06T10:16:50.550832 #1] ERROR -- Kontena::WebsocketClient: unable to connect to server: Connection refused - connect(2) for 192.168.66.1:9292
I, [2017-10-06T10:16:50.565558 #1]  INFO -- Kontena::WebsocketClient: reconnect attempt 5 in 14.55s...
I, [2017-10-06T10:17:05.128235 #1]  INFO -- Kontena::WebsocketClient: connecting to master at ws://192.168.66.1:9292
E, [2017-10-06T10:17:05.140768 #1] ERROR -- Kontena::WebsocketClient: unable to connect to server: Connection refused - connect(2) for 192.168.66.1:9292
I, [2017-10-06T10:17:05.147431 #1]  INFO -- Kontena::WebsocketClient: reconnect attempt 6 in 2.91s...
I, [2017-10-06T10:17:08.059366 #1]  INFO -- Kontena::WebsocketClient: connecting to master at ws://192.168.66.1:9292
E, [2017-10-06T10:17:08.074719 #1] ERROR -- Kontena::WebsocketClient: unable to connect to server: Connection refused - connect(2) for 192.168.66.1:9292
I, [2017-10-06T10:17:08.075968 #1]  INFO -- Kontena::WebsocketClient: reconnect attempt 7 in 57.74s...
I, [2017-10-06T10:18:05.816037 #1]  INFO -- Kontena::WebsocketClient: connecting to master at ws://192.168.66.1:9292
I, [2017-10-06T10:18:05.867017 #1]  INFO -- Kontena::WebsocketClient: unsecure connection established without SSL
```

## TODO
- [ ] specs

### CODE REVIEW

**Comment by SpComb** [comment]
Could also consider implementing the reconnect logic within the `Kontena::Websocket::Client`, which would simplify the actor code in both the server and agent.

The reconnect logic within the `Kontena::Websocket::Client` could be smarter for things like `socket.kontena.io` that resolve to multiple addresses. Looking at how the server `Cloud::WebsocketClient` behaves, it tends to try reconnecting to the same IP multiple times in a row, which makes the round-robin DNS less effective.

**Comment by SpComb** [comment]
Sticking `status/design-review` on here because I'm not planning on fixing the specs until the design makes sense.

**Comment by SpComb** [comment]
Inferring that the design is fine, and starting to fix the specs to test the implicit `start!` -> `connect!` -> `reconnect! / `connected! -> `disconnected!` -> `reconnect!` -> `connect!` state machine in the agent and server, rather than implementing the reconnect logic in the kontena-websocket-client library.

**Comment by SpComb** [comment]
@nevalla WDYT about the server `CloudWebsocketClient` reconnect timings? Each worker uses randomized exponential backoff with an initial delay of 0-5s, and that goes up to 0-120s, so the workers should spread their reconnect attempts across a long enough time period, which grows longer the longer the connections are offline:

```
I, [2017-10-30T14:53:46.418170 #12]  INFO -- Cloud::WebsocketClient: cloud connection closed with code 1006: read timeout after 4.983617513s while waiting 5.0s for pong
I, [2017-10-30T14:53:46.418887 #12]  INFO -- Cloud::WebsocketClient: reconnect attempt 1 in 4.59s...
I, [2017-10-30T14:53:47.287174 #20]  INFO -- Cloud::WebsocketClient: cloud connection closed with code 1006: read timeout after 4.999686016s while waiting 5.0s for pong
I, [2017-10-30T14:53:47.288032 #20]  INFO -- Cloud::WebsocketClient: reconnect attempt 1 in 0.21s...
I, [2017-10-30T14:53:47.494245 #20]  INFO -- Cloud::WebsocketClient: Connecting to cloud at wss://socket.kontena.io/platform
I, [2017-10-30T14:53:47.945408 #8]  INFO -- Cloud::WebsocketClient: cloud connection closed with code 1006: read timeout after 4.999259182s while waiting 5.0s for pong
I, [2017-10-30T14:53:47.948964 #8]  INFO -- Cloud::WebsocketClient: reconnect attempt 1 in 3.57s...
I, [2017-10-30T14:53:48.386137 #16]  INFO -- Cloud::WebsocketClient: cloud connection closed with code 1006: read timeout after 4.999533491s while waiting 5.0s for pong
I, [2017-10-30T14:53:48.386774 #16]  INFO -- Cloud::WebsocketClient: reconnect attempt 1 in 2.64s...

I, [2017-10-30T14:53:47.494245 #20]  INFO -- Cloud::WebsocketClient: Connecting to cloud at wss://socket.kontena.io/platform
I, [2017-10-30T14:53:51.011752 #12]  INFO -- Cloud::WebsocketClient: Connecting to cloud at wss://socket.kontena.io/platform
I, [2017-10-30T14:53:51.029096 #16]  INFO -- Cloud::WebsocketClient: Connecting to cloud at wss://socket.kontena.io/platform
I, [2017-10-30T14:53:51.522516 #8]  INFO -- Cloud::WebsocketClient: Connecting to cloud at wss://socket.kontena.io/platform

...

I, [2017-10-30T14:56:29.013094 #8]  INFO -- Cloud::WebsocketClient: Connecting to cloud at wss://socket.kontena.io/platform
I, [2017-10-30T14:56:29.837238 #8]  INFO -- Cloud::WebsocketClient: Connected to wss://socket.kontena.io/platform with ssl verify: /OU=Domain Control Validated/OU=PositiveSSL Wildcard/CN=*.kontena.io (issuer /C=GB/ST=Greater Manchester/L=Salford/O=COMODO CA Limited/CN=COMODO RSA Domain Validation Secure Server CA)
I, [2017-10-30T14:56:30.247797 #20]  INFO -- Cloud::WebsocketClient: Connecting to cloud at wss://socket.kontena.io/platform
I, [2017-10-30T14:56:31.074249 #20]  INFO -- Cloud::WebsocketClient: Connected to wss://socket.kontena.io/platform with ssl verify: /OU=Domain Control Validated/OU=PositiveSSL Wildcard/CN=*.kontena.io (issuer /C=GB/ST=Greater Manchester/L=Salford/O=COMODO CA Limited/CN=COMODO RSA Domain Validation Secure Server CA)
I, [2017-10-30T14:57:29.729743 #16]  INFO -- Cloud::WebsocketClient: Connecting to cloud at wss://socket.kontena.io/platform
I, [2017-10-30T14:57:30.709683 #16]  INFO -- Cloud::WebsocketClient: Connected to wss://socket.kontena.io/platform with ssl verify: /OU=Domain Control Validated/OU=PositiveSSL Wildcard/CN=*.kontena.io (issuer /C=GB/ST=Greater Manchester/L=Salford/O=COMODO CA Limited/CN=COMODO RSA Domain Validation Secure Server CA)
I, [2017-10-30T14:57:33.600650 #12]  INFO -- Cloud::WebsocketClient: Connecting to cloud at wss://socket.kontena.io/platform
I, [2017-10-30T14:57:34.441702 #12]  INFO -- Cloud::WebsocketClient: Connected to wss://socket.kontena.io/platform with ssl verify: /OU=Domain Control Validated/OU=PositiveSSL Wildcard/CN=*.kontena.io (issuer /C=GB/ST=Greater Manchester/L=Salford/O=COMODO CA Limited/CN=COMODO RSA Domain Validation Secure Server CA)
```

In this example with four workers, the connections went down at 14:53:46 - 14:53:48... the first reconnects happened at 14:53:47 - 14:53:51, and then ~3min later when the connection comes back up, the reconnects happen at 14:56:29 - 14:57:33.

File: agent/lib/kontena/websocket_client.rb [hunk 1]
```diff
@@ -74,14 +75,33 @@ def rpc_client
...
-    def connect!
-      @connecting = true
+    def reconnect!
+      if @reconnect_attempt > 16
+        backoff = RECONNECT_BACKOFF
+      else
+        backoff = [RECONNECT_BACKOFF, CONNECT_INTERVAL * 2 ** @reconnect_attempt].min
+      end
+
+      backoff *= rand
```
Comment (jakolehm) [inline agent/lib/kontena/websocket_client.rb]: Maybe add comment here why rand (jitter)?

Comment (SpComb) [reply agent/lib/kontena/websocket_client.rb]: Refactored as a separate `reconnect_backoff` function with doc comments

**Pull request merged**